### PR TITLE
Reland "[SimplifyCFG] Check if the return instruction causes undefined behavior"

### LINKED ISSUE #60717: In a function with a `noundef` return value, replace `ret undef` with `unreachable`
`noundef` on the return value means that

> If the value representation contains any undefined or poison bits, the behavior is undefined.

so if the function `ret`urns `undef`, that must be [`unreachable`](https://llvm.org/docs/LangRef.html#unreachable-instruction).

Alive2 confirms this is allowed: <https://alive2.llvm.org/ce/z/bW2DoM>
```
define i32 @src() noundef {
%start:
  ret i32 undef
}
=>
define i32 @tgt() noundef {
%start:
  assume i1 0
}
Transformation seems to be correct!
```

But seemingly it doesn't happen today: <https://llvm.godbolt.org/z/79a91o59q>

### PR DESCRIPTION
**Pull request by dianqk**
Fixes #60717.
#69498 has been merged.

### CODE REVIEW

**Comment by DianQK** [comment]
Hmm, it looks like we're going to create `release/18`, so let me merge it first. If we find new issues, we can revert it.

**Review by nikic — APPROVED** [review]
LGTM, but please wait until after LLVM 18 has branched to merge it. Please also perform the merge manually to avoid squashing the commits.

**Comment by DianQK** [comment]
> LGTM, but please wait until after LLVM 18 has branched to merge it. Please also perform the merge manually to avoid squashing the commits.

Okay, waiting until LLVM 19 is fine. Yes, I'm going to rebase and push.

**Pull request merged**
